lib/auth/machineid/machineidv1/spiffe_federation_syncer.go

```diff
@@ -0,0 +1,582 @@
+// syncer.
+type SPIFFEFederationSyncerConfig struct {
+	Backend       backend.Backend
+	Store         spiffeFederationStore
+	EventsWatcher eventsWatcher
+	Logger        *slog.Logger
+	Clock         clockwork.Clock
+
+	// MinSyncInterval is the minimum interval between syncs. If an upstream
+	// trust domain specifies a refresh hint that is less than this value, this
+	// value will be used instead. This allows us to prevent a poorly configured
+	// upstream trust domain from causing excessive load on the local cluster.
+	MinSyncInterval time.Duration
+	// MaxSyncInterval is the maximum interval between syncs. If an upstream
+	// trust domain specifies a refresh hint that is greater than this value,
+	// this value will be used instead. This allows us to prevent a poorly
+	// configured upstream trust domain from causing excessive staleness in the
+	// local cluster.
+	MaxSyncInterval time.Duration
+	// DefaultSyncInterval is the interval between syncs that will be used if an
+	// upstream trust domain does not specify a refresh hint.
+	DefaultSyncInterval time.Duration
+
+	// SyncTimeout is the maximum time that a sync operation is allowed to take.
+	// If a sync operation takes longer than this value, it will be aborted and
+	// retried.
+	SyncTimeout time.Duration
+
+	SPIFFEFetchOptions []federation.FetchOption
+}
+
+// SPIFFEFederationSyncer is a syncer that manages the trust bundles of
+// federated clusters. It runs on a single elected auth server.
+type SPIFFEFederationSyncer struct {
+	cfg SPIFFEFederationSyncerConfig
+}
+
+const (
+	minRefreshInterval     = time.Minute * 1
+	maxRefreshInterval     = time.Hour * 24
+	defaultRefreshInterval = time.Minute * 5
+	defaultSyncTimeout     = time.Second * 30
+)
+
+// NewSPIFFEFederationSyncer creates a new SPIFFEFederationSyncer.
+func NewSPIFFEFederationSyncer(cfg SPIFFEFederationSyncerConfig) (*SPIFFEFederationSyncer, error) {
+	switch {
+	case cfg.Backend == nil:
+		return nil, trace.BadParameter("backend: must be non-nil")
+	case cfg.Store == nil:
+		return nil, trace.BadParameter("store: must be non-nil")
+	case cfg.Logger == nil:
+		return nil, trace.BadParameter("logger: must be non-nil")
+	case cfg.Clock == nil:
+		return nil, trace.BadParameter("clock: must be non-nil")
+	}
+	if cfg.MinSyncInterval == 0 {
+		cfg.MinSyncInterval = minRefreshInterval
+	}
+	if cfg.MaxSyncInterval == 0 {
+		cfg.MaxSyncInterval = maxRefreshInterval
+	}
+	if cfg.DefaultSyncInterval == 0 {
+		cfg.DefaultSyncInterval = defaultRefreshInterval
+	}
+	if cfg.SyncTimeout == 0 {
+		cfg.SyncTimeout = defaultSyncTimeout
+	}
+	return &SPIFFEFederationSyncer{
+		cfg: cfg,
+	}, nil
+}
+
+func (s *SPIFFEFederationSyncer) Run(ctx context.Context) error {
+	// TODO: Should this go into a loop?
+	return backend.RunWhileLocked(ctx, backend.RunWhileLockedConfig{
+		// TODO: Evaluate sensible TTL/retry
+		LockConfiguration: backend.LockConfiguration{
```

Comment:
TODO: Speak to TimR or similar on deciding sensible values for here and whether this should go into its own retry loop.

Reply:
Related comment

https://github.com/gravitational/teleport/pull/45586#discussion_r1727423225

I feel like the logic can be simplify by relying on watcher reset where all collection is refetched instead of implementing periodic sync in loop `syncTrustDomainLoop`